test/regtest.py

```diff
@@ -104,43 +104,142 @@ def metadata(file):
 
-        print '  [%.2fs]' % round(elapsed, 2)
-        sys.stdout.flush()
-  
-except KeyboardInterrupt:
-  pass
+        str_result += '  [%.2fs]' % round(elapsed, 2)
 
-print
-print ' PASSED count:', passed
-print ' FAILED count:', failed
+        # log the info for this test
+        logging.info(str_result)
 
-if timeouts > 0:
-  print 'TIMEOUT count:', timeouts
-
-if unknowns > 0:
-  print 'UNKNOWN count:', unknowns
+        # returns a status code based on the results of the test
+        if passed:
+            return 0
+        elif timedout:
+            return 1
+        elif unknown:
+            return 2
+        else:
+            return -1
+
+
+def init_worker():
+    """
+    Set up the worker processes to ignore SIGINT altogether. This confines all
+    the cleanup code to the parent process.
+
+    Reference: http://stackoverflow.com/a/6191991/3342427
+    """
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+
+def main():
+    """
+    Main entry point for the test suite.
+    """
+    # configure the CLI
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--exhaustive", help="be exhaustive", action="store_true")
+    parser.add_argument("--threads", action="store", dest="n_threads", default=1, type=int,
+                        help="execute regressions using the selected number of threads in parallel")
+    parser.add_argument("--log", action="store", dest="log_level", default="DEBUG", type=str,
+                        help="sets the logging level (DEBUG, INFO, WARNING)")
+    parser.add_argument("--output_log", action="store", dest="log_path", type=str,
+                        help="sets the output log path. (std out by default)")
+    args = parser.parse_args()
+
+
+    # configure the logging
+    log_format = ''
+    log_level = logging.DEBUG
+
+    # add more log levels later (if needed)
+    if args.log_level.upper() == "INFO":
+        log_level = logging.INFO
+    elif args.log_level.upper() == "WARNING":
+        log_level = logging.WARNING
+
+    # if the user supplied a log path, write the logs to that file.
+    # otherwise, write the logs to std out.
+    if args.log_path:
+        logging.basicConfig(filename=args.log_path, format=log_format, level=log_level)
+    else:
+        logging.basicConfig(format=log_format, level=log_level)
+
+    logging.debug("Creating Pool with '%d' Workers" % args.n_threads)
+    p = multiprocessing.Pool(processes=args.n_threads, initializer=init_worker)
+
+    # start the tests
+    logging.info("Running regression tests...")
+    passed = failed = timeouts = unknowns = 0
+
+    # start processing the tests. this blocks the main thread
+    async_results = []
+    for test in glob.glob("./**/*.c") if args.exhaustive else glob.glob("./basic/*.c"):
+        async_results.append(p.apply_async(process_test, args=(test, args,)))
+
+    try:
+        # keep the main thread active while there are active workers
+        count = 0
+        while True:
```

Comment:
Do we really need to do “polling” here, repeatedly checking the results of each async task?

It seems like it would be cleaner simply to tally the results once the tests have completed — either “naturally” or via an interrupt. In other words, after the call to `p.join()`.